ls1intum/Artemis
# [Feature] Add static analysis for C programming exercises

### Checklist
- [ ]  Documentation: I extended the Artemis Documentation with infos about the feature

### Motivation and Context



### Description

The operating system course always had a dedicated test case for the GCC static analysis feature. It was not especially readable since it could contain a lot of different errors that could not be related to static analysis. Furthermore, due to the length of the test output the messages got even cut off in some cases, hence not being helpful for students when debugging their code. This PR integrates the static analysis for GCC into Artemis and provides useful default categories, that enable instructors to make only SCA related output visible for students. 

In order to support readability I did some small changes to the UI. Firstly, I made the feedback-text Monospace, which closes #3948. Additionally, I cut off the preview of the feedback after the first line. From my perspective it looked kinda confusing before if multiple lines get squashed into a single line. I added screenshots below, so you can see how it looks. What do you think?

I am looking forward for your feedback. I just extended the existing SCA tests for C where applicable. Basically the ones that were also used for the [Swift SCA PR](https://github.com/ls1intum/Artemis/pull/2512). Should I do more testing or this fine already?

Also, is Jenkins support optional or required?

### Steps for Testing

Before you can start testing the issue you need to get the Bamboo Plugin ready. To make life easier, here is a already compiled plugin that contains the updated version of the [static-code-analysis-parser](https://github.com/ls1intum/static-code-analysis-parser/pull/11). If you want to compile the plugin yourself, refer to the info in the linked PR.

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a C Programming Exercise with SCA enabled
4. Clone the student repository to your local machine
5. Edit e.g. hello_world.c and try to cause a SCA violation (For example add a double free())
6. Push the changes and wait for the build to finish
7. Open the Feedback section next to "Refresh files" and see that an entry was added with "CODE ISSUE" (See screenshots)
8. Look at the categories of the SCA results. Try to disable one that was seen in the Feedback and rerun the build.
9. Check if the correct category got filtered out. 

### Review Progress





- Code Review
 
- Manual Tests

### Test Coverage



StaticCodeAnalysisConfigurer.java: 100%
BambooBuildPlanService.java: 90%
BambooService.java: 85%

### Screenshots


This is how the output looks in monospace right now:
![image](https://user-images.githubusercontent.com/30267166/131231542-d7e53515-9ad6-47dd-8c53-0e9dc15809c3.png)

The before mentioned line break:
![image](https://user-images.githubusercontent.com/30267166/131231178-8eb4ac2e-46cf-47d3-885d-8b9492ecb132.png)

All these categories should be visible:
![image](https://user-images.githubusercontent.com/30267166/131231211-6bc1e6ef-ce37-4237-bce9-d917cb81a7ec.png)
